cranelift/wasm/src/state/module_state.rs

```diff
@@ -21,11 +24,46 @@ pub struct ModuleTranslationState {
     pub(crate) wasm_types: WasmTypes,
 }
 
+fn cranelift_to_wasmparser_type(ty: Type) -> WasmResult<wasmparser::Type> {
+    Ok(match ty {
+        types::I32 => wasmparser::Type::I32,
+        types::I64 => wasmparser::Type::I64,
+        types::F32 => wasmparser::Type::F32,
+        types::F64 => wasmparser::Type::F64,
+        _ => {
+            return Err(WasmError::Unsupported(format!(
+                "Cannot convert Cranelift type to Wasm signature: {:?}",
+                ty
+            )));
+        }
+    })
+}
+
 impl ModuleTranslationState {
     /// Creates a new empty ModuleTranslationState.
     pub fn new() -> Self {
         Self {
             wasm_types: PrimaryMap::new(),
         }
     }
+
+    /// Create a new ModuleTranslationState with the given function signatures,
+    /// provided in terms of Cranelift types. The provided slice of signatures
+    /// is indexed by signature number, and contains pairs of (args, results)
+    /// slices.
+    pub fn from_func_sigs(sigs: &[(Box<[Type]>, Box<[Type]>)]) -> WasmResult<Self> {
+        let mut wasm_types = PrimaryMap::with_capacity(sigs.len());
+        for &(ref args, ref results) in sigs {
+            let args: Vec<wasmparser::Type> = args
+                .iter()
+                .map(|&ty| cranelift_to_wasmparser_type(ty))
+                .collect::<Result<_, _>>()?;
```

Comment:
(am i understanding correctly that there's Something that infers that we want a `Result<Vec<T>>` from a `Vec<Result<T>>`? I can imagine how this works, but I didn't know it was available, this is brilliant!)

Reply:
Indeed, there is apparently a `FromIterator` implementation for `Result` that makes this work!